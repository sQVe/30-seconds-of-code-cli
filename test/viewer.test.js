const { head } = require('../src/helpers');
const { logSnippet, prettyPrint } = require('../src/viewer');

const layouts = [...Array.from('idce'), 'idce'];
const snippet = {
  code: 'const head = arr => arr[0];',
  example: 'head([1, 2, 3]); // 1',
  id: 'head',
  description: [
    'Returns the head of a list.',
    'Use `arr[0]` to return the first element of the passed array.',
  ].join('\n'),
};

describe('Viewer', () => {
  describe('prettyPrint()', () => {
    it('should pretty print based on given snippet', () => {
      expect(
        prettyPrint(
          Object.keys(snippet).map(k => ({ [k]: snippet[k] }))
        ).toString()
      ).toMatchSnapshot();
      expect(prettyPrint([snippet])).toMatchSnapshot();
    });
  });

  describe('logSnippet()', () => {
    const stringifySpy = jest.spyOn(JSON, 'stringify');
    global.console = { ...global.console, log: jest.fn() };

    afterEach(() => {
      console.log.mockClear();
      stringifySpy.mockClear();
    });

    it('should handle snippets in a array', () => {
      logSnippet({ layout: 'ic' }, [snippet]);

      expect(head(console.log.mock.calls)).toMatchSnapshot();
    });

    it('should handle snippet in object', () => {
      logSnippet({ layout: 'ic' }, snippet);

      expect(head(console.log.mock.calls)).toMatchSnapshot();
    });

    layouts.forEach(layout => {
      it('should log JSON given json flag', () => {
        logSnippet({ layout, json: true }, snippet);

        expect(head(console.log.mock.calls)).toMatchSnapshot();
      });

      it('should log prettyPrint() by default', () => {
        logSnippet({ layout }, snippet);
        expect(head(console.log.mock.calls)).toMatchSnapshot();
      });
    });
  });
});
