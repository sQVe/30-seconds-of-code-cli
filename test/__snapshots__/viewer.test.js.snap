// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewer logSnippet() should handle snippet in object 1`] = `
Array [
  "[35m[1mhead[22m[39m
[35m[1m----[22m[39m
[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];",
]
`;

exports[`Viewer logSnippet() should handle snippets in a array 1`] = `
Array [
  "[35m[1mhead[22m[39m
[35m[1m----[22m[39m
[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];",
]
`;

exports[`Viewer logSnippet() should log JSON given json flag 1`] = `
Array [
  "[{\\"id\\":\\"head\\"}]",
]
`;

exports[`Viewer logSnippet() should log JSON given json flag 2`] = `
Array [
  "[{\\"description\\":\\"Returns the head of a list.\\\\nUse \`arr[0]\` to return the first element of the passed array.\\"}]",
]
`;

exports[`Viewer logSnippet() should log JSON given json flag 3`] = `
Array [
  "[{\\"code\\":\\"const head = arr => arr[0];\\"}]",
]
`;

exports[`Viewer logSnippet() should log JSON given json flag 4`] = `
Array [
  "[{\\"example\\":\\"head([1, 2, 3]); // 1\\"}]",
]
`;

exports[`Viewer logSnippet() should log JSON given json flag 5`] = `
Array [
  "[{\\"id\\":\\"head\\",\\"description\\":\\"Returns the head of a list.\\\\nUse \`arr[0]\` to return the first element of the passed array.\\",\\"code\\":\\"const head = arr => arr[0];\\",\\"example\\":\\"head([1, 2, 3]); // 1\\"}]",
]
`;

exports[`Viewer logSnippet() should log prettyPrint() by default 1`] = `
Array [
  "[35m[1mhead[22m[39m
[35m[1m----[22m[39m",
]
`;

exports[`Viewer logSnippet() should log prettyPrint() by default 2`] = `
Array [
  "[0mReturns the head of a list.[0m
[0mUse [33marr[0][39m to return the first element of the passed array.[0m",
]
`;

exports[`Viewer logSnippet() should log prettyPrint() by default 3`] = `
Array [
  "[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];",
]
`;

exports[`Viewer logSnippet() should log prettyPrint() by default 4`] = `
Array [
  "head([[32m1[39m, [32m2[39m, [32m3[39m]); [32m// 1[39m",
]
`;

exports[`Viewer logSnippet() should log prettyPrint() by default 5`] = `
Array [
  "[35m[1mhead[22m[39m
[35m[1m----[22m[39m
[0mReturns the head of a list.[0m
[0mUse [33marr[0][39m to return the first element of the passed array.[0m

[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];

head([[32m1[39m, [32m2[39m, [32m3[39m]); [32m// 1[39m",
]
`;

exports[`Viewer prettyPrint() should pretty print based on given snippet 1`] = `
"[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];

head([[32m1[39m, [32m2[39m, [32m3[39m]); [32m// 1[39m

[35m[1mhead[22m[39m
[35m[1m----[22m[39m

[0mReturns the head of a list.[0m
[0mUse [33marr[0][39m to return the first element of the passed array.[0m"
`;

exports[`Viewer prettyPrint() should pretty print based on given snippet 2`] = `
"[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];

head([[32m1[39m, [32m2[39m, [32m3[39m]); [32m// 1[39m

[35m[1mhead[22m[39m
[35m[1m----[22m[39m
[0mReturns the head of a list.[0m
[0mUse [33marr[0][39m to return the first element of the passed array.[0m"
`;
